MetaDataset.py

```diff
@@ -885,23 +883,142 @@ def init_seq_order(self, epoch=None, seq_list=None):
+      # We have to re-calculate the seq_idx's because we will sort the datasets below.
+      self.dataset_sorted_seq_idx_list = []
+      counters = [0] * len(self.datasets)
+      for seq_idx in seq_order:
+        dataset_idx, _ = self._seq_idx_to_dataset_seq_idx(seq_idx)
+        dataset_seq_idx = counters[dataset_idx]
+        counters[dataset_idx] += 1
+        self.dataset_sorted_seq_idx_list.append((dataset_idx, dataset_seq_idx))
+
+      # We only want to load those sequences in the sub-datasets that appear in seq_order. For this, we extract
+      # sequence lists containing the subset of sequences for each dataset from seq_order.
+      seq_lists = [[] for _ in self.datasets]
+      for seq_idx in seq_order:
+        dataset_idx, dataset_seq_idx = self._seq_idx_to_dataset_seq_idx(seq_idx)
+        dataset = self.datasets[self.dataset_idx2key_map[dataset_idx]]
+        seq_tag = dataset.get_tag(dataset_seq_idx)
+        seq_lists[dataset_idx].append(seq_tag)
+
+      # It may be large, so better delete it early, we don't need it anymore.
+      del seq_order
+
+      # Re-initialize sequence orders of sub-datasets with created sequence list.
+      for dataset_idx, dataset_key in self.dataset_idx2key_map.items():
+        self.datasets[dataset_key].init_seq_order(epoch=epoch, seq_list=seq_lists[dataset_idx])
+
+    else:
+      self.dataset_sorted_seq_idx_list = []  # We will fill this as we go
+      self.used_num_seqs_per_subset = [0] * len(self.datasets)
 
     return True
 
+  def _create_dataset_seq_idx_boundaries(self):
+    """
+    Creates a list storing the positions where the different sub-datasets start within the full sequence list.
+    This happens to be the cumulative sum of the number of sequences in the sub-datasets.
+
+    :rtype: list[int]
+    """
+    dataset_seq_idx_boundaries = [0]
+    for dataset_num_seqs in self.used_num_seqs_per_subset:
+      dataset_seq_idx_boundaries.append(dataset_num_seqs + dataset_seq_idx_boundaries[-1])
+
+    return dataset_seq_idx_boundaries
+
+  def _seq_idx_to_dataset_seq_idx(self, seq_idx):
+    """
+    Maps the sequence index (before sorting) to a dataset and a corresponding sequence index within this dataset.
+
+    :param int seq_idx:
+    :rtype: (int,int)
+    """
+    assert seq_idx < self.dataset_seq_idx_boundaries[-1]
+
+    dataset_idx = 0
+    while self.dataset_seq_idx_boundaries[dataset_idx + 1] <= seq_idx:
+      dataset_idx += 1
+
+    dataset_seq_idx = seq_idx - self.dataset_seq_idx_boundaries[dataset_idx]
+
+    return dataset_idx, dataset_seq_idx
+
+  def _get_random_dataset_seq_order(self, total_num_seqs):
+    """
+    Choose datasets randomly but preserve order within each dataset. This sorting method is unique to CombinedDataset.
+
+    :param int total_num_seqs:
+    :param int epoch:
+    :returns: sequence order
+    :rtype: list[int]
+    """
+    full_epoch = (self.epoch - 1) // self.partition_epoch + 1 if self.epoch else 1
```

Comment:
Btw, there is also `self._get_random_seed_for_epoch`, which might be better to use here.

Reply:
Ah, yes, looks better.